internal/webhook/inferenceservice/mutating.go

```diff
@@ -0,0 +1,212 @@
+
+	default: // this should not enter, but keep it just for sanity check if allowedTypes is updated.
+		log.Info("Unknown connection type, skipping injection", "connectionType", connectionType)
+		return false, nil, nil
+	}
+}
+
+// injectOCIImagePullSecrets injects imagePullSecrets into spec.predictor.imagePullSecrets for OCI connections.
+func (w *ConnectionWebhook) injectOCIImagePullSecrets(obj *unstructured.Unstructured, secretName string) error {
+	imagePullSecrets, found, err := unstructured.NestedSlice(obj.Object, IsvcConfigs.ImagePullSecretPath...)
+	if err != nil {
+		return fmt.Errorf("failed to get spec.predictor.imagePullSecrets: %w", err)
+	}
+	// did not have imagePullSecrets (upon CREATE), just set to the new secret
+	if !found {
+		imagePullSecrets = []interface{}{
+			map[string]interface{}{
+				"name": secretName,
+			},
+		}
+		return unstructured.SetNestedSlice(obj.Object, imagePullSecrets, IsvcConfigs.ImagePullSecretPath...)
+	}
+
+	// if already some secrets(upon UPDATE), and the secret is already there, fast exit
+	for _, secret := range imagePullSecrets {
+		if secretMap, ok := secret.(map[string]interface{}); ok {
+			if name, exists := secretMap["name"]; exists && name == secretName {
+				return nil
+			}
+		}
+	}
+
+	// add new secret to the secrets(upon UPDATE)
+	newImagePullSecret := map[string]interface{}{
+		"name": secretName,
+	}
+	imagePullSecrets = append(imagePullSecrets, newImagePullSecret)
+
+	return unstructured.SetNestedSlice(obj.Object, imagePullSecrets, IsvcConfigs.ImagePullSecretPath...)
+}
+
+// injectURIStorageUri injects storageUri into spec.predictor.model.storageUri for URI connections.
+func (w *ConnectionWebhook) injectURIStorageUri(obj *unstructured.Unstructured, secret *corev1.Secret) error {
+	var storageUri string
+	if uri, exists := secret.Data["URI"]; exists {
+		// The secret data is already base64 decoded by Kubernetes, so we can use it directly
+		storageUri = string(uri)
+	} else {
+		return errors.New("secret does not contain 'URI' data key")
+	}
+
+	// Set the storageUri directly
+	return unstructured.SetNestedField(obj.Object, storageUri, "spec", "predictor", "model", "storageUri")
+}
+
+// injectS3StorageKey injects storage key into spec.predictor.model.storage.key for S3 connections.
+func (w *ConnectionWebhook) injectS3StorageKey(obj *unstructured.Unstructured, secretName string) error {
+	model, found, err := unstructured.NestedMap(obj.Object, IsvcConfigs.ModelPath...)
+	if err != nil {
+		return fmt.Errorf("failed to get spec.predictor.model: %w", err)
+	}
+	if !found {
+		return errors.New("found no spec.predictor.model set in resource")
+	}
+
+	storageMap, err := webhookutils.GetOrCreateNestedMap(model, "storage")
+	if err != nil {
+		return fmt.Errorf("failed to get or create nested map for storage: %w", err)
+	}
+	storageMap["key"] = secretName
+	model["storage"] = storageMap
+
+	return unstructured.SetNestedMap(obj.Object, model, IsvcConfigs.ModelPath...)
```

Comment:
nit: maybe the process of get/amend/set for nested map and slices can be improved by having an utility method that accept a mutator function ?